File: en/codes/swift/chapter_graph/graph_adjacency_list.swift
/**
 * File: graph_adjacency_list.swift
 * Created Time: 2023-02-01
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Undirected graph class based on adjacency list */
public class GraphAdjList {
    // Adjacency list, key: vertex, value: all adjacent vertices of that vertex
    public private(set) var adjList: [Vertex: [Vertex]]

    /* Constructor */
    public init(edges: [[Vertex]]) {
        adjList = [:]
        // Add all vertices and edges
        for edge in edges {
            addVertex(vet: edge[0])
            addVertex(vet: edge[1])
            addEdge(vet1: edge[0], vet2: edge[1])
        }
    }

    /* Get the number of vertices */
    public func size() -> Int {
        adjList.count
    }

    /* Add edge */
    public func addEdge(vet1: Vertex, vet2: Vertex) {
        if adjList[vet1] == nil || adjList[vet2] == nil || vet1 == vet2 {
            fatalError("Parameter error")
        }
        // Add edge vet1 - vet2
        adjList[vet1]?.append(vet2)
        adjList[vet2]?.append(vet1)
    }

    /* Remove edge */
    public func removeEdge(vet1: Vertex, vet2: Vertex) {
        if adjList[vet1] == nil || adjList[vet2] == nil || vet1 == vet2 {
            fatalError("Parameter error")
        }
        // Remove edge vet1 - vet2
        adjList[vet1]?.removeAll { $0 == vet2 }
        adjList[vet2]?.removeAll { $0 == vet1 }
    }

    /* Add vertex */
    public func addVertex(vet: Vertex) {
        if adjList[vet] != nil {
            return
        }
        // Add a new linked list to the adjacency list
        adjList[vet] = []
    }

    /* Remove vertex */
    public func removeVertex(vet: Vertex) {
        if adjList[vet] == nil {
            fatalError("Parameter error")
        }
        // Remove the vertex vet's corresponding linked list from the adjacency list
        adjList.removeValue(forKey: vet)
        // Traverse other vertices' linked lists, removing all edges containing vet
        for key in adjList.keys {
            adjList[key]?.removeAll { $0 == vet }
        }
    }

    /* Print the adjacency list */
    public func print() {
        Swift.print("Adjacency list =")
        for (vertex, list) in adjList {
            let list = list.map { $0.val }
            Swift.print("\(vertex.val): \(list),")
        }
    }
}

#if !TARGET

@main
enum GraphAdjacencyList {
    /* Driver Code */
    static func main() {
        /* Initialize undirected graph */
        let v = Vertex.valsToVets(vals: [1, 3, 2, 5, 4])
        let edges = [[v[0], v[1]], [v[0], v[3]], [v[1], v[2]], [v[2], v[3]], [v[2], v[4]], [v[3], v[4]]]
        let graph = GraphAdjList(edges: edges)
        print("\nAfter initialization, the graph is")
        graph.print()

        /* Add edge */
        // Vertices 1, 2 i.e., v[0], v[2]
        graph.addEdge(vet1: v[0], vet2: v[2])
        print("\nAfter adding edge 1-2, the graph is")
        graph.print()

        /* Remove edge */
        // Vertices 1, 3 i.e., v[0], v[1]
        graph.removeEdge(vet1: v[0], vet2: v[1])
        print("\nAfter removing edge 1-3, the graph is")
        graph.print()

        /* Add vertex */
        let v5 = Vertex(val: 6)
        graph.addVertex(vet: v5)
        print("\nAfter adding vertex 6, the graph is")
        graph.print()

        /* Remove vertex */
        // Vertex 3 i.e., v[1]
        graph.removeVertex(vet: v[1])
        print("\nAfter removing vertex 3, the graph is")
        graph.print()
    }
}

#endif

File: en/codes/javascript/chapter_graph/graph_adjacency_list.js
/**
 * File: graph_adjacency_list.js
 * Created Time: 2023-02-09
 * Author: Justin (xiefahit@gmail.com)
 */

const { Vertex } = require('../modules/Vertex');

/* Undirected graph class based on adjacency list */
class GraphAdjList {
    // Adjacency list, key: vertex, value: all adjacent vertices of that vertex
    adjList;

    /* Constructor */
    constructor(edges) {
        this.adjList = new Map();
        // Add all vertices and edges
        for (const edge of edges) {
            this.addVertex(edge[0]);
            this.addVertex(edge[1]);
            this.addEdge(edge[0], edge[1]);
        }
    }

    /* Get the number of vertices */
    size() {
        return this.adjList.size;
    }

    /* Add edge */
    addEdge(vet1, vet2) {
        if (
            !this.adjList.has(vet1) ||
            !this.adjList.has(vet2) ||
            vet1 === vet2
        ) {
            throw new Error('Illegal Argument Exception');
        }
        // Add edge vet1 - vet2
        this.adjList.get(vet1).push(vet2);
        this.adjList.get(vet2).push(vet1);
    }

    /* Remove edge */
    removeEdge(vet1, vet2) {
        if (
            !this.adjList.has(vet1) ||
            !this.adjList.has(vet2) ||
            vet1 === vet2
        ) {
            throw new Error('Illegal Argument Exception');
        }
        // Remove edge vet1 - vet2
        this.adjList.get(vet1).splice(this.adjList.get(vet1).indexOf(vet2), 1);
        this.adjList.get(vet2).splice(this.adjList.get(vet2).indexOf(vet1), 1);
    }

    /* Add vertex */
    addVertex(vet) {
        if (this.adjList.has(vet)) return;
        // Add a new linked list to the adjacency list
        this.adjList.set(vet, []);
    }

    /* Remove vertex */
    removeVertex(vet) {
        if (!this.adjList.has(vet)) {
            throw new Error('Illegal Argument Exception');
        }
        // Remove the vertex vet's corresponding linked list from the adjacency list
        this.adjList.delete(vet);
        // Traverse other vertices' linked lists, removing all edges containing vet
        for (const set of this.adjList.values()) {
            const index = set.indexOf(vet);
            if (index > -1) {
                set.splice(index, 1);
            }
        }
    }

    /* Print the adjacency list */
    print() {
        console.log('Adjacency list =');
        for (const [key, value] of this.adjList) {
            const tmp = [];
            for (const vertex of value) {
                tmp.push(vertex.val);
            }
            console.log(key.val + ': ' + tmp.join());
        }
    }
}

if (require.main === module) {
    /* Driver Code */
    /* Initialize undirected graph */
    const v0 = new Vertex(1),
        v1 = new Vertex(3),
        v2 = new Vertex(2),
        v3 = new Vertex(5),
        v4 = new Vertex(4);
    const edges = [
        [v0, v1],
        [v1, v2],
        [v2, v3],
        [v0, v3],
        [v2, v4],
        [v3, v4],
    ];
    const graph = new GraphAdjList(edges);
    console.log('\nAfter initialization, the graph is');
    graph.print();

    /* Add edge */
    // Vertices 1, 2 are v0, v2
    graph.addEdge(v0, v2);
    console.log('\nAfter adding edge 1-2, the graph is');
    graph.print();

    /* Remove edge */
    // Vertices 1, 3, are v0, v1
    graph.removeEdge(v0, v1);
    console.log('\nAfter removing edge 1-3, the graph is');
    graph.print();

    /* Add vertex */
    const v5 = new Vertex(6);
    graph.addVertex(v5);
    console.log('\nAfter adding vertex 6, the graph is');
    graph.print();

    /* Remove vertex */
    // Vertex 3 is v1
    graph.removeVertex(v1);
    console.log('\nAfter removing vertex 3, the graph is');
    graph.print();
}

module.exports = {
    GraphAdjList,
};
